src/commands.cpp

```diff
@@ -324,6 +324,30 @@ CommandPutFile::CommandPutFile(MegaClient* client, TransferSlot* ctslot, int ms)
     arg("v", 2);
     arg("s", tslot->fa->size);
     arg("ms", ms);
+
+    // send minimum set of different tree's roots for API to check overquota
+    set<handle> targetRoots;
+    beginarray("t");
+    for (auto &file : tslot->transfer->files)
+    {
+        if (!ISUNDEF(file->h))
+        {
+            Node *node = client->nodebyhandle(file->h);
+            if (node)
+            {
+                handle rootnode = client->getrootnode(node)->nodehandle;
```

Comment:
@mattw-mega do you think we should avoid to scale the tree up to the rootnode several times for each file? It is done here, but also when the transfer fails. We could save whether the `Transfer::file` is a foreign one or not by adding a `bool isForeign` to the `File` object. We would also need to restore the status upon serialization, or add it to the cached data.
I've checked and only `rename()` seems to use the `getrootnode()`, despite I guess there are plenty of code going upwards to the rootnode for multiple purposes, so probably it is not worthy. If you feel the same way, I think we can merge

Reply:
I think that's a very minor cost, I wouldn't worry about it unless it shows up when profiling in some scenario.  I think it's fine to merge (I did spot one other thing which I'll comment, but it's not a showstopper, could be adjusted anytime later on)